## HYPOTHESIS TESTING ##

# MAGEZI RICHARD ELIJAH
## M24B13/013
### B27258


* An assumption or idea, specifically a statistical claim about an unknown parameter.
* Presumptive statement

* Defining Hypotheses
* Null hypothesis (H0): In statistics, the null hypothesis is a general statement or default position that there is no relationship/no effects/no difference between two measured cases or among groups. In other words, it is a basic assumption or made based on the problem knowledge.
* Example using diamonds dataset: There is no relationship between carat and price

* Alternative hypothesis (H1/Ha/HA): The alternative hypothesis is contrary to the null hypothesis. 
* Example: There is a relationship between carat and price.

        ## Key Terms of Hypothesis Testing

* Level of significance(alpha): It refers to the degree of significance in which we accept or reject the null hypothesis. 100% accuracy is not possible for accepting a hypothesis, so we, therefore, select a level of significance that is usually 5%. 

* P-value: The P value, or calculated probability, is the probability of finding the observed/extreme results when the null hypothesis(H0) of a study-given problem is true. If your P-value is less than the chosen significance level then you reject the null hypothesis i.e. accept that your sample claims to support the alternative hypothesis.

* Test Statistic: The test statistic is a numerical value calculated from sample data during a hypothesis test, used to determine whether to reject the null hypothesis. It is compared to a critical value or p-value to make decisions about the statistical significance of the observed results.

* Critical value: The critical value in statistics is a threshold or cutoff point used to determine whether to reject the null hypothesis in a hypothesis test.
* Degrees of freedom: Degrees of freedom are associated with the variability or freedom one has in estimating a parameter. The degrees of freedom are related to the sample size and determine the shape.

       ## Steps in hypothesis testing
* 1. State and define null and alternative hypotheses
* 2. Choose a significance level (usually alpha = 0.05 or 5%)
* 3. Collect and analyse data
* 4. Calculate the test statistic and corresponding p-value
* 5. Make conclusions

In [3]:
#libraries
import pandas as pd
import numpy as np
import scipy.stats as stats

#Import datasets
diamonds=pd.read_excel("diamonds_new.xlsx")
diamonds

price  carat        cut color clarity  depth  table     x     y     z
0        326   0.23      Ideal     E     SI2   61.5   55.0  3.95  3.98  2.43
1        326   0.21    Premium     E     SI1   59.8   61.0  3.89  3.84  2.31
2        327   0.23       Good     E     VS1   56.9   65.0  4.05  4.07  2.31
3        334   0.29    Premium     I     VS2   62.4   58.0  4.20  4.23  2.63
4        335   0.31       Good     J     SI2   63.3   58.0  4.34  4.35  2.75
...      ...    ...        ...   ...     ...    ...    ...   ...   ...   ...
53915   2757   0.72      Ideal     D     SI1   60.8   57.0  5.75  5.76  3.50
53916   2757   0.72       Good     D     SI1   63.1   55.0  5.69  5.75  3.61
53917   2757   0.70  Very Good     D     SI1   62.8   60.0  5.66  5.68  3.56
53918   2757   0.86    Premium     H     SI2   61.0   58.0  6.15  6.12  3.74
53919   2757   0.75      Ideal     D     SI2   62.2   55.0  5.83  5.87  3.64

[53920 rows x 10 columns]

## Statistical tests to investigate

**A. T-test**
* Compares the mean of two different samples or groups when the sample size is ≤ 30 
* Data should follow normal or Gaussian distribution
* Different types of T-tests:

**i. One Sample T-test:** 
* Used to compare the mean value of a sample with a known mean or average. 
* The null and alternative hypotheses of this T-test are as follows:
* i. Null Hypothesis: μ = M
* ii. Alternative Hypothesis: μ ≠ M
* Where μ is the sample mean and M is the known mean/population mean.

In [5]:
#Let's generate a random dataset that meets the above assumptions
import random
# We will use random data for the samples and label them as 'a' and 'b' respectively
a = [random.gauss(50, 20) for x in range(30)] #a variable with a list of 50
b = [random.gauss(55, 15) for x in range(30)]

#If we assume the population mean is 40
#Null hypothesis is the sample mean = population mean(40)
#alpha=0.05
#Therefore, p<0.05 (reject null hypothesis)
#Let's test it for sample a
t_stat, p_value = stats.ttest_1samp(a, 40, axis=0)

#Print the t-stat and the p-value
print("T-statistic:", t_stat)
print("P-value:", p_value)
print(np.mean(a))

T-statistic: 3.323912287864602
P-value: 0.00241310214420894
50.19250857400527


*      Interpretation of the results: 
* The p-value is less than 0.05(signifance level) and the sample mean(50.19) is also ≠ population mean(40).
* We reject the null hypothesis and accept the alternative hypothesis

**ii. Independent Sample T-test or Two Sample T-test**
* Null hypothesis: The means of samples are equal

In [7]:
# Null hypothesis: mean of a = mean of b
t_stat, p_value = stats.ttest_ind(a,b,equal_var = False)
#print results
print("T-statistic:", t_stat)
print("P-value:", p_value)
print(np.mean(a), np.mean(b))

#We can also add an if/else function which is defined as such:  
if p_value < 0.05:      
    print(" We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis")  

T-statistic: -0.42146151000303733
P-value: 0.6750024151067231
50.19250857400527 52.15236256714354
We can accept the null hypothesis


*      Interpretation of the results: 
* The p-value is greater than 0.05(signifance level) and the mean of a = mean of b
* We accept the null hypothesis and reject the alternative hypothesis

**iii. Dependent or Paired Sample T-test**
* A substantial difference between two related variables is tested using a univariate test
*  Null hypothesis: mean of b - mean of a = 0

In [8]:
t_stat,p_value = stats.ttest_rel(a,b)
#print results
print("T-statistic:", t_stat)
print("P-value:", p_value)
print(np.mean(a) - np.mean(b))
#Print conclusion
if p_value < 0.05:      
    print(" We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis") 

T-statistic: -0.4337456828470809
P-value: 0.6676828455083401
-1.9598539931382746
We can accept the null hypothesis


**B. Z-test**
* Used to compare the difference between a sample and a known population mean or between the means of two independent samples. 
* Sample size is ≥ 30 and 
* The population standard deviation is known.
* There are two different types of Z-test:
* One-sample Z-test: Used to compare one sample mean with a known population mean.
i. Null Hypothesis: μ = μ₀
ii. Alternative Hypothesis: μ ≠ μ₀
Where μ is the population mean and μ₀ is a specified value you compare with the population mean.

**i. One-sample Z-test**
* Compares ones sample mean with a known population mean
* Null hypothesis: μ = μ₀
* Alternative Hypothesis: μ ≠ μ₀
* Where μ is the population mean and μ₀ is a specified value you compare with the population mean.

In [13]:
#Import the z-test library
from statsmodels.stats.weightstats import ztest as ztest
# Let's use the diamonds dataset
# Let's assume the mean price of a diamond is 200 and the standard deviation 40
#Specified value = 200
#Standard deviation = 40
# Range=53920 (testing hypothesis for the 53920 diamonds in the dataset)
diamonds['price'] = [random.gauss(200, 40) for x in range(53920)]

# One Sample Z-test for sample 'a'
# Null Hypothesis: Mean of diamond price = 200 
z_stat, p_value = ztest(diamonds['price'], value=200)

#print results
print("Z-statistic:", z_stat)
print("P-value:", p_value)
print(np.mean(diamonds['price']))

#print conclusion
if p_value < 0.05:      
    print(" We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis") 

Z-statistic: 0.8059089168995973
P-value: 0.4202953693934157
200.1390453783929
We can accept the null hypothesis


*     Interpretation of results:
* We see that the mean price of the diamonds within the dataset is equal to the assumed mean price of 200

**ii. Two Sample Z-Test**
* Examines two independent data sets to determine whether or not the sample means of the two groups are identical.
* Null Hypothesis: Mean of a = mean of b
* Similar to the independent t-test
* To carry this out you have to first check if the variables are independent
* So let's use the Cancer dataset (medical dataframe) as the variable radius_mean amongst two different groups: patients with malignant and benign tumors (M and B)
* Null hypothesis: Mean of radius_mean of malignant tumors = Mean of radius_mean of benign tumors

In [16]:
#Import libraries
from scipy import stats  
from statsmodels.stats import weightstats as stests  

#State null hypothesis:
#Mean of radius_mean of malignant tumors = Mean of radius_mean of benign tumors
#

#Specify the groups
M=medical[medical["diagnosis"] == "M"].radius_mean
B=medical[medical["diagnosis"] == "B"].radius_mean

In [17]:
#Run the two-sided z-test
z_test ,p_value = stests.ztest(M, x2 = B, value = 0, alternative = 'two-sided')  

#Print results
print("Z-statistic:", z_stat)
print("P-value:", p_value)
print(np.mean(M), np.mean(B)) 

#Print conclusion 
if p_value < 0.05:  
    print("We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis")  

Z-statistic: 0.8059089168995973
P-value: 1.0131691292988755e-142
17.462830188679245 12.146523809523808
We can reject the null hypothesis


*       Interpretation of results
* Null hypothesis is rejected
* Therefore, the mean of the radius_means of the tumors amongst Benign and Malignat groups are not equal

## Further Reading ##
* Hypothesis testing for Analysis of variance, or ANOVA or F-test
* Hypothesis testing for Chisquare test
* Hypothesis testing for Mann-Whitney U Test
* Hypothesis testing for Shapiro-wilk's test
* Hypothesis testing for Wilcoxon Signed-Rank Test
* Hypothesis testing for Kruskal-Wallis H Test
* Hypothesis testing for Kolmogorov–Smirnov test
* Differences between parametric and non-parametric tests



## Practice Exercise ##
* Using the diamonds dataset, test the hypothesis that average price of all carat categories are the same.

# Hypothesis 1
## Average price of all table categories are the same.

In [6]:
print(diamonds['table'].unique())

[55.  61.  65.  58.  57.  56.  54.  62.  59.  63.  60.  55.3 69.  64.
 56.4 52.  55.8 66.  70.  68.  67.  54.4 53.  57.1 53.6 54.8 54.2 54.9
 56.7 53.9 57.8 54.1 55.1 51.  58.1 56.9 56.5 53.7 60.7 56.2 56.6 56.1
 56.3 57.2 57.5 60.9 59.8 62.3 60.1 55.9 50.1 62.2 54.5 55.7 53.4 54.7
 58.8 59.4 62.5 54.3 60.6 55.2 58.6 49.  50.  54.6 62.6 57.7 56.8 57.9
 60.3 55.6 59.5 57.4 62.4 60.4 53.3 60.2 60.8 58.5 57.6 43.  59.9 63.5
 57.3 61.4 58.3 55.4 58.2 53.5 59.7 55.5 58.7 53.8 59.1 53.2 59.3 61.1
 51.6 95.  61.2 59.2 60.5 58.4 53.1 52.8 44.  61.3 61.9 61.8 59.6 61.5
 58.9 64.3 62.8 71.  52.4 63.3 61.7 64.2 73.  65.4 63.4 62.1 79.  76.
 61.6]


In [14]:
#Testing the hypothesis that the average price of all carat categories are the same
#Import scipy.stats
import scipy.stats

#Grouping the data using carat categories
groups=[group['price'].values for name, group in diamonds.groupby("table")]

#ANOVA
h_statistic, p_value_h1 =stats.kruskal(*groups)
print(f'H-statistics also known as Krushal_Wallis H-test:{h_statistic},p_value: {p_value_h1}')

#Print conclusion 
if p_value < 0.05:  
    print("We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis")  

H-statistics also known as Krushal_Wallis H-test::1798.3451261087168,p_value: 1.464467845021479e-293
We can reject the null hypothesis


Null hypotheis is rejected which means that the statistical evidence shows that the data is inconsitent with the null hypothesis. The data shows that the pricing structure vaies based on the table category

# Hypothesis 2
## Average price of Premuim diamonds being 150

In [8]:
#Filtering out premium diamonds
pre_diamonds=diamonds[diamonds["cut"]=="Premium"]
pre_diamonds

price  carat      cut color clarity  depth  table     x     y     z
1        326   0.21  Premium     E     SI1   59.8   61.0  3.89  3.84  2.31
3        334   0.29  Premium     I     VS2   62.4   58.0  4.20  4.23  2.63
12       345   0.20  Premium     E     SI2   60.2   62.0  3.79  3.75  2.27
13       345   0.32  Premium     E      I1   60.9   58.0  4.38  4.42  2.68
23       355   0.24  Premium     I     VS1   62.5   57.0  3.97  3.94  2.47
...      ...    ...      ...   ...     ...    ...    ...   ...   ...   ...
53908   2756   0.79  Premium     E     SI2   61.4   58.0  6.03  5.96  3.68
53910   2756   0.71  Premium     E     SI1   60.5   55.0  5.79  5.74  3.49
53911   2756   0.71  Premium     F     SI1   59.8   62.0  5.74  5.73  3.43
53914   2757   0.72  Premium     D     SI1   62.7   59.0  5.69  5.73  3.58
53918   2757   0.86  Premium     H     SI2   61.0   58.0  6.15  6.12  3.74

[13789 rows x 10 columns]

In [13]:
t_statistic, p_value=stats.ttest_1samp(pre_diamonds["price"],150)

#The print statement
print(f"T_statistic: {t_statistic},p_value: {p_value}")

#Print conclusion 
if p_value < 0.05:  
    print("We can reject the null hypothesis")  
else:  
    print("We can accept the null hypothesis")  

T_statistic: 119.73104883785925,p_value: 0.0
We can reject the null hypothesis


The null hypothesis is rejected meaning that the average price of Premium diamonds is not equal to 150